### PR TITLE
DATACOUCH-515 add to list of query dsl queries to n1ql

### PR DESCRIPTION
**Pull request by mikereiche**
This Pull Request is so that @daschl can follow the changes.

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@mikereiche Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-data-couchbase&pullRequestId=212)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-data-couchbase&pullRequestId=212) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@mikereiche Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-data-couchbase&pullRequestId=212)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-data-couchbase&pullRequestId=212) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@mikereiche Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-data-couchbase&pullRequestId=212)!

File: src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java [hunk 1]
```diff
@@ -52,77 +66,177 @@ public QueryCriteria or(String key) {
 		return new QueryCriteria(this.criteriaChain, key, null, ChainOperator.OR);
 	}
 
+	public QueryCriteria child(QueryCriteria criteria) {
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java:69]: does the `child` naming align with similar implementations in other spring-data projects?

File: src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java [hunk 2]
```diff
@@ -52,77 +66,177 @@ public QueryCriteria or(String key) {
...
 
 	public QueryCriteria gte(@Nullable Object o) {
 		operator = ">=";
-		value = o;
+		value = new Object[]{o};
+		return this;
+	}
+
+	public QueryCriteria startingWith(@Nullable Object o) {
+		regex("^"+o.toString());
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java:115]: wouldn't that be more the LIKE operator?

File: src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java [hunk 3]
```diff
@@ -52,77 +66,177 @@ public QueryCriteria or(String key) {
...
+		return this;
+	}
+
+	public QueryCriteria endingWith(@Nullable Object o) {
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java:119]: same as the LIKE ?

File: src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java [hunk 4]
```diff
@@ -52,77 +66,177 @@ public QueryCriteria or(String key) {
...
+
+	public QueryCriteria notcontaining(@Nullable Object o) {
+		value=new QueryCriteria[]{child(containing(o))};
+		operator ="NOT";
+		format = format = "%2$s( %3$s )";
+		chainOperator = ChainOperator.NOT;
+		return (QueryCriteria)this;
+	}
+
+	public QueryCriteria like(@Nullable Object o) {
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java:146]: do a real LIKE ?

File: src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java [hunk 5]
```diff
@@ -133,6 +247,26 @@ private String maybeWrapValue(Object value) {
...
+		v[0]=fieldName;
+		v[1]=operator;
+		for(int i=0; i< value.length;i++){
+			if(value[i] instanceof QueryCriteria)
+				v[i+2]="("+((QueryCriteria)value[i]).export()+")";
+			else
+				v[i+2]=maybeWrapValue(value[i]);
+		}
+
+		if(format == null)
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java:263]: let's use { } always, even for single lines. It makes it much less error prone if refactoed later.

File: src/main/java/org/springframework/data/couchbase/repository/query/N1qlQueryCreator.java [hunk 6]
```diff
@@ -60,18 +60,55 @@ private QueryCriteria from(final Part part, final CouchbasePersistentProperty pr
 														 final Iterator<Object> parameters) {
 
 		final Part.Type type = part.getType();
-
+/*
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/repository/query/N1qlQueryCreator.java:63]: these are still missing?

File: src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java [hunk 7]
```diff
@@ -52,77 +66,177 @@ public QueryCriteria or(String key) {
...
+	}
+
+	public QueryCriteria containing(@Nullable Object o) {
+		operator = "CONTAINS";
+		value = new Object[]{o};
+		format = "%2$s (%1$s,   %3$s)";
+		return this;
+	}
+
+	public QueryCriteria notcontaining(@Nullable Object o) {
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java:138]: please use method names like notContaining

File: src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java [hunk 8]
```diff
@@ -52,77 +66,177 @@ public QueryCriteria or(String key) {
...
+	}
+
+	public QueryCriteria like(@Nullable Object o) {
+		operator = "REGEX_CONTAINS";
+		value=new Object[]{o};// might want to transform from regex to 'like' (i.e. .* -> %, . -> ? )
+		format = "%2$s (%1$s,   %3$s)";
+		return this;
+	}
+
+	public QueryCriteria notlike(@Nullable Object o) {
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java:153]: notLike etc.

File: src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java [hunk 9]
```diff
@@ -52,77 +66,177 @@ public QueryCriteria or(String key) {
...
+		value=new QueryCriteria[]{child(like(o))};
```
Comment (daschl) [inline src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java:154]: formatting is off across the class

File: src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java (hunk 2 above)
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java:121]: LIKE '<value>%' should also work.

File: src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java (hunk 3 above)
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java:125]: LIKE '%arg' would also work

File: src/main/java/org/springframework/data/couchbase/repository/query/N1qlQueryCreator.java (hunk 6 above)
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/repository/query/N1qlQueryCreator.java:63]: Just two left now - NEAR and WITHIN.
WITHIN is straight-forward.  There doesn't appear to be a N1QL equivalent to NEAR.

File: src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java (hunk 1 above)
Comment (mikereiche) [reply src/main/java/org/springframework/data/couchbase/core/query/QueryCriteria.java]: Agree - child is not a good name.  I've changed it to 'wrap' and also made it private.

**Pull request closed without merging**